File: sdks/python/src/memoturn/anthropic.py
"""Drop-in Anthropic wrapper — records ``messages.create`` calls (including streaming)
as generations."""
from __future__ import annotations

import json
from typing import Any, Optional

from ._stream import _ABANDONED, _RecordingStream, _get
from .client import Memoturn, Trace
from .decorator import get_client

#: Model parameters copied onto the generation (allowlist — everything else is dropped).
_PARAM_KEYS = ("max_tokens", "temperature", "top_p", "top_k", "stop_sequences")


def wrap_anthropic(client: Any, memoturn: Optional[Memoturn] = None, *, trace: Optional[Trace] = None) -> Any:
    """Patch ``client.messages.create`` to trace calls. Returns the same client.

    Streaming calls (``stream=True``) are recorded too: the returned stream is wrapped so
    chunks are forwarded unchanged to the caller while being accumulated into the same
    output/usage shape as a non-streaming call, and the generation is closed when the
    stream is exhausted, errors, or is abandoned (closed/garbage-collected/idle-timed-out
    before completion).
    """
    mt = memoturn or get_client()
    _patch_messages(client, mt, trace)
    return client


def _patch_messages(client: Any, mt: Memoturn, trace: Optional[Trace]) -> None:
    messages = client.messages
    original = messages.create

    def create(*args: Any, **kwargs: Any) -> Any:
        t = trace or mt.trace(name="anthropic.messages")
        system = kwargs.get("system")
        msgs = kwargs.get("messages")
        gen = t.generation(
            name="anthropic.messages",
            model=kwargs.get("model"),
            provider="anthropic",
            # Keep the system prompt alongside the messages when present.
            input={"system": system, "messages": msgs} if system is not None else msgs,
            modelParameters={k: v for k, v in kwargs.items() if k in _PARAM_KEYS},
        )

        if kwargs.get("stream"):
            return _stream_messages(original, args, kwargs, gen)

        try:
            resp = original(*args, **kwargs)
            gen.end(output=_content(resp), usage=_usage(resp))
            return resp
        except Exception as e:  # noqa: BLE001
            gen.end(level="ERROR", statusMessage=str(e))
            raise

    messages.create = create  # type: ignore[assignment]


def _stream_messages(original: Any, args: tuple, kwargs: dict, gen: Any) -> Any:
    try:
        raw = original(*args, **kwargs)
    except Exception as e:  # noqa: BLE001
        gen.end(level="ERROR", statusMessage=str(e))
        raise

    acc = _AnthropicAccumulator()

    def on_chunk(event: Any) -> None:
        acc.add(event)

    def on_done(err: Any) -> None:
        output, usage = acc.finalize()
        if err is _ABANDONED:
            gen.end(level="WARNING", statusMessage="stream ended before completion", output=output, usage=usage)
        elif err is not None:
            gen.end(level="ERROR", statusMessage=str(err), output=output, usage=usage)
        else:
            gen.end(output=output, usage=usage)

    return _RecordingStream(raw, on_chunk, on_done)


class _AnthropicAccumulator:
    """Accumulates ``messages.create`` streaming events (``message_start`` /
    ``content_block_*`` / ``message_delta``) into the same ``(output, usage)`` shape the
    non-streaming path builds."""

    def __init__(self) -> None:
        self._blocks: dict[int, dict[str, Any]] = {}
        self._json_buffers: dict[int, str] = {}
        self._input_tokens: Optional[int] = None
        self._output_tokens: Optional[int] = None
        self._cache_read_tokens: Optional[int] = None
        self._cache_creation_tokens: Optional[int] = None

    def add(self, event: Any) -> None:
        etype = _get(event, "type")
        if etype == "message_start":
            self._on_message_start(event)
        elif etype == "content_block_start":
            self._on_block_start(event)
        elif etype == "content_block_delta":
            self._on_block_delta(event)
        elif etype == "content_block_stop":
            self._on_block_stop(event)
        elif etype == "message_delta":
            self._on_message_delta(event)

    def _on_message_start(self, event: Any) -> None:
        message = _get(event, "message")
        usage = _get(message, "usage") if message is not None else None
        if usage is None:
            return
        self._input_tokens = _get(usage, "input_tokens")
        self._cache_read_tokens = _get(usage, "cache_read_input_tokens")
        self._cache_creation_tokens = _get(usage, "cache_creation_input_tokens")

    def _on_block_start(self, event: Any) -> None:
        index = _get(event, "index", 0)
        block = _get(event, "content_block")
        init: dict[str, Any] = {}
        if block is not None:
            btype = _get(block, "type")
            if btype is not None:
                init["type"] = btype
            for key in ("id", "name"):
                val = _get(block, key)
                if val is not None:
                    init[key] = val
            if btype == "text":
                init["text"] = _get(block, "text") or ""
            if btype == "thinking":
                init["thinking"] = _get(block, "thinking") or ""
        self._blocks[index] = init

    def _on_block_delta(self, event: Any) -> None:
        index = _get(event, "index", 0)
        delta = _get(event, "delta")
        block = self._blocks.setdefault(index, {})
        dtype = _get(delta, "type") if delta is not None else None
        if dtype == "text_delta":
            block.setdefault("type", "text")
            block["text"] = (block.get("text") or "") + (_get(delta, "text") or "")
        elif dtype == "input_json_delta":
            block.setdefault("type", "tool_use")
            self._json_buffers[index] = self._json_buffers.get(index, "") + (_get(delta, "partial_json") or "")
        elif dtype == "thinking_delta":
            block.setdefault("type", "thinking")
            block["thinking"] = (block.get("thinking") or "") + (_get(delta, "thinking") or "")
        elif dtype == "signature_delta":
            block["signature"] = _get(delta, "signature")

    def _on_block_stop(self, event: Any) -> None:
        index = _get(event, "index", 0)
        buf = self._json_buffers.get(index)
        if buf is None:
            return
        block = self._blocks.setdefault(index, {})
        try:
            block["input"] = json.loads(buf)
        except json.JSONDecodeError:
            block["input"] = buf

    def _on_message_delta(self, event: Any) -> None:
        usage = _get(event, "usage")
        if usage is None:
            return
        out_tokens = _get(usage, "output_tokens")
        if out_tokens is not None:
            self._output_tokens = out_tokens

    def finalize(self) -> tuple[Any, Optional[dict[str, Any]]]:
        output = [self._blocks[i] for i in sorted(self._blocks)] if self._blocks else None
        if self._input_tokens is None and self._output_tokens is None:
            usage = None
        else:
            usage = {
                "promptTokens": self._input_tokens,
                "completionTokens": self._output_tokens,
                "totalTokens": (self._input_tokens or 0) + (self._output_tokens or 0),
                "cacheReadTokens": self._cache_read_tokens,
                "cacheCreationTokens": self._cache_creation_tokens,
            }
        return output, usage


def _content(resp: Any) -> Any:
    blocks = getattr(resp, "content", None)
    if blocks is None:
        return str(resp)
    try:
        return [b.model_dump() if hasattr(b, "model_dump") else b for b in blocks]
    except Exception:  # noqa: BLE001
        return str(blocks)


def _usage(resp: Any) -> Optional[dict[str, Any]]:
    usage = getattr(resp, "usage", None)
    if usage is None:
        return None
    in_tokens = getattr(usage, "input_tokens", None)
    out_tokens = getattr(usage, "output_tokens", None)
    return {
        "promptTokens": in_tokens,
        "completionTokens": out_tokens,
        "totalTokens": (in_tokens or 0) + (out_tokens or 0),
        "cacheReadTokens": getattr(usage, "cache_read_input_tokens", None),
        "cacheCreationTokens": getattr(usage, "cache_creation_input_tokens", None),
    }

File: sdks/js/src/anthropic.ts
import type { Memoturn, MemoturnTrace } from "./client.js";
import { tapStream } from "./stream.js";

/** Model parameters worth recording — everything else (messages, system, tools, …) is payload. */
const MODEL_PARAMETER_ALLOWLIST = ["max_tokens", "temperature", "top_p", "top_k", "stop_sequences"] as const;

/**
 * Drop-in wrapper for the Anthropic SDK. Wraps `client.messages.create` so each call is
 * recorded as a memoturn generation (model, params, usage incl. cache tokens, latency, output)
 * — including streaming calls (`stream: true`), which are accumulated into the same
 * output/usage shape as a non-streaming call while still being yielded to the caller in real
 * time (no buffering, no added latency).
 *
 *   const anthropic = wrapAnthropic(new Anthropic(), memoturn);
 *   await anthropic.messages.create({ model, max_tokens, messages });
 *
 * Pass `{ trace }` to nest generations under an existing trace; otherwise each call
 * gets its own trace. Pass `{ streamTimeoutMs }` to override the idle-stream abandonment
 * backstop (default 120s).
 */
export function wrapAnthropic<T extends object>(
  client: T,
  memoturn: Memoturn,
  options: { trace?: MemoturnTrace; traceName?: string; streamTimeoutMs?: number } = {},
): T {
  return new Proxy(client, {
    get(target, prop, receiver) {
      const value = Reflect.get(target, prop, receiver);
      if (prop === "messages") return wrapMessages(value, memoturn, options);
      return value;
    },
  });
}

/** Map Anthropic usage (incl. prompt-cache tokens) to the recorded usage shape — shared by
 * the non-streaming path and the stream accumulator's `.finalize()`. */
function mapAnthropicUsage(usage: any): Record<string, number> | undefined {
  if (!usage) return undefined;
  return {
    promptTokens: usage.input_tokens,
    completionTokens: usage.output_tokens,
    totalTokens: (usage.input_tokens ?? 0) + (usage.output_tokens ?? 0),
    ...(usage.cache_read_input_tokens != null ? { cacheReadTokens: usage.cache_read_input_tokens } : {}),
    ...(usage.cache_creation_input_tokens != null ? { cacheCreationTokens: usage.cache_creation_input_tokens } : {}),
  };
}

/** Merge Anthropic streaming events (`message_start`/`content_block_*`/`message_delta`) into
 * the same `{ output, usage }` shape the non-streaming path records. */
function createAnthropicStreamAccumulator() {
  const blocks: Record<number, any> = {};
  const jsonBuffers: Record<number, string> = {};
  let usage: any;

  return {
    add(event: any) {
      switch (event?.type) {
        case "message_start": {
          if (event.message?.usage) usage = { ...usage, ...event.message.usage };
          break;
        }
        case "content_block_start": {
          blocks[event.index] = { ...event.content_block };
          break;
        }
        case "content_block_delta": {
          const block = (blocks[event.index] ??= {});
          const delta = event.delta ?? {};
          if (delta.type === "text_delta") {
            block.text = (block.text ?? "") + (delta.text ?? "");
          } else if (delta.type === "input_json_delta") {
            jsonBuffers[event.index] = (jsonBuffers[event.index] ?? "") + (delta.partial_json ?? "");
          } else if (delta.type === "thinking_delta") {
            block.thinking = (block.thinking ?? "") + (delta.thinking ?? "");
          } else if (delta.type === "signature_delta") {
            block.signature = delta.signature;
          } else {
            // Unknown/future delta type — merge whatever fields it carries as a best effort.
            for (const [k, v] of Object.entries(delta)) {
              if (k !== "type") block[k] = v;
            }
          }
          break;
        }
        case "content_block_stop": {
          const buffered = jsonBuffers[event.index];
          if (buffered !== undefined) {
            const block = (blocks[event.index] ??= {});
            try {
              block.input = JSON.parse(buffered);
            } catch {
              // Malformed/partial JSON — never throw out of the accumulator.
              block.input = buffered;
            }
            delete jsonBuffers[event.index];
          }
          break;
        }
        case "message_delta": {
          if (event.usage) usage = { ...usage, ...event.usage };
          break;
        }
        default:
          break; // message_stop and anything else: stream exhaustion is what tapStream detects.
      }
    },
    finalize() {
      const output = Object.keys(blocks)
        .map(Number)
        .sort((a, b) => a - b)
        .map((i) => blocks[i])
        .filter(Boolean);
      return { output, usage: mapAnthropicUsage(usage) };
    },
  };
}

function wrapMessages(
  messages: any,
  memoturn: Memoturn,
  options: { trace?: MemoturnTrace; traceName?: string; streamTimeoutMs?: number },
): any {
  return new Proxy(messages, {
    get(target, prop, receiver) {
      const original = Reflect.get(target, prop, receiver);
      if (prop !== "create" || typeof original !== "function") return original;

      return async function create(params: any, ...rest: any[]) {
        const trace = options.trace ?? memoturn.trace({ name: options.traceName ?? "anthropic.messages" });
        const modelParameters: Record<string, unknown> = {};
        for (const key of MODEL_PARAMETER_ALLOWLIST) {
          if (params?.[key] !== undefined) modelParameters[key] = params[key];
        }
        const generation = trace.generation({
          name: "anthropic.messages",
          model: params?.model,
          provider: "anthropic",
          modelParameters,
          // Keep `system` (system prompt) alongside the messages when present.
          input: params?.system != null ? { system: params.system, messages: params?.messages } : params?.messages,
        });

        if (params?.stream) {
          try {
            // Anthropic's streaming protocol always includes final usage in `message_delta`
            // — no `stream_options`-equivalent injection is needed to get usage.
            const response = await original.call(target, params, ...rest);
            const accumulator = createAnthropicStreamAccumulator();
            return tapStream(
              response,
              {
                onChunk: (event: any) => accumulator.add(event),
                onDone: (err, reason) => {
                  if (reason === "error") {
                    generation.end({ level: "ERROR", statusMessage: String(err), ...accumulator.finalize() });
                  } else if (reason === "abandoned") {
                    generation.end({
                      level: "WARNING",
                      statusMessage: "stream ended before completion",
                      ...accumulator.finalize(),
                    });
                  } else {
                    generation.end({ ...accumulator.finalize() });
                  }
                },
              },
              { idleTimeoutMs: options.streamTimeoutMs },
            );
          } catch (err) {
            generation.end({ level: "ERROR", statusMessage: String(err) });
            throw err;
          }
        }

        try {
          const response = await original.call(target, params, ...rest);
          generation.end({ output: response?.content ?? response, usage: mapAnthropicUsage(response?.usage) });
          return response;
        } catch (err) {
          generation.end({ level: "ERROR", statusMessage: String(err) });
          throw err;
        }
      };
    },
  });
}
